### ISSUE TITLE
[Feature Request] Spell checking for OCR result

### ISSUE THREAD

**Issue by heinrichI** [dmMaze/BallonsTranslator#591]
Can you add dictionary check of OCR results? It helps to find many errors quickly. This function works well for subtitle recognition in [Subtitle Edit](https://github.com/SubtitleEdit/subtitleedit). There it is done through via [Libre Office dictionaries](https://extensions.libreoffice.org/?Tags%5B%5D=50)
![image](https://github.com/user-attachments/assets/f601bbb2-b13c-43f6-9c16-73f6a2297452)

**Comment by bropines** [comment]
We discussed this once. QT doesn't seem to have a proper spell checking tool. In theory, there may be something in the vastness of libraries... But maybe DmMaze can tell you something

**Comment by heinrichI** [comment]
Why support from QT? Scan the OCR results with a dictionary, remove the names and show the remaining non-dictionary words in a separate panel.

**Comment by bropines** [comment]
> Why support from QT? Scan the OCR results with a dictionary, remove the names and show the remaining non-dictionary words in a separate panel.

That is, you are suggesting that on the way from OCR we should also add a dictionary that will obviously not predict some words absolutely accurately?

**Comment by heinrichI** [comment]
> That is, you are suggesting that on the way from OCR we should also add a dictionary that will obviously not predict some words absolutely accurately?

Yes, add a dictionary. Of course, it will not know all the words, but and OCR is not 100% accurate. This is just a simplification of text proofreading, so that you don't have to reread everything. If a word is not in the dictionary, then either it is an unknown word and we add it to the user dictionary, or it is a name and we add it to the temporary list of names, or it is an OCR error and you need to edit the text. The screenshot from the Subtitle Edit shows the buttons for adding to dictionaries, the editing window, possible correct options or autoreplacement.

**Comment by vanderalex** [comment]
I also think that dictionary will be very useful, please, think of implementing this function. In my translating, usually there are 3-5 errors on the page after OCR, and if comics has handwriting text... you understand :) And it's very hard to localize these mistakes sometimes...

**Comment by heinrichI** [comment]
I made a small prototype using spells.hunspell, can anyone help me finalize it and implement it into BallonsTranslator? https://github.com/heinrichI/BallonsTranslator/commit/4126d6bd7ad3e13430e262deea1d3b9f3ce0c599

![Image](https://github.com/user-attachments/assets/7e0a1b0a-a741-4b0b-b2da-33ab10681184)

**Comment by bropines** [comment]
Напишу на двух языках сразу. 
1. Прототип слишком награможденный. 
2. Проще использовать https://pyenchant.github.io/pyenchant потому что автор не светит везде что он поддерживает когото, и потому что эта библиотека в разы удобнее по api. Так же я написал в PR что абсолютно бессмысленно создавать отдельное окно, Умные люди придумали контекстные меню. Текст можно спокойно подсвечивать, через методы qt. 

I will write in two languages ​​at once. 
1. The prototype is too award. 
2. It is easier to use https://pyenchant.github.io/pyenchant because the author does not shine everywhere that he supports someone, and because this library is much more convenient in the API. I also wrote in PR that it is absolutely pointless to create a separate window, smart people came up with contextual menus. The text can be calmly illuminated through QT methods.

**Comment by heinrichI** [comment]
> is absolutely pointless to create a separate window,

Why highlight after translation, correct and re-translate, if it is more logical to check errors after the OCR and only then translate?

**Comment by bropines** [comment]
> > is absolutely pointless to create a separate window,
> 
> Why highlight after translation, correct and re-translate, if it is more logical to check errors after the OCR and only then translate?

Well, if I understand correctly, during the recognition, this window flies out, and says "here is the word, it is strange, let's make the correction options for you." Provided that our application is created for automatic recognition and subsequent processing. If it is convenient for you to reduce every mistake, in the top, but in the case when you need to drive out 10 projects in a row, and for example I need it, because I translate the Yonkoma (small comics) then it interferes. And here, as a very convenient solution, a stupid context menu is suitable. I have the implementation of such a solution, but the problem is that I am stupid and I can’t understand normally Qt.

https://github.com/user-attachments/assets/60403122-589d-442e-a98c-5cc7941463e9

**Comment by heinrichI** [comment]
Yes, a pop-up window will completely remove the automatism. But you still need to reread for errors, or is it assumed that the OCR will work completely automatically and without errors? Then at least a panel with a list of unrecognized words is needed, so that you can see them right away, and not reread all the phrases, looking for underlining.

**Comment by bropines** [comment]
@dmmaze question. What implementation do you like more? We just discussed personally, and we want to know your opinion about the implementation of Spellchecker

**Comment by heinrichI** [comment]
@bropines Can you upload your branch with the dictionary in the text panels?

**Comment by bropines** [comment]
> [@bropines](https://github.com/bropines) Can you upload your branch with the dictionary in the text panels?

She is not ready yet. And I can't find time to finish her.

**Comment by heinrichI** [comment]
Made a version with a panel:

https://github.com/user-attachments/assets/44389a41-7d0e-4586-822b-7f6d8fcd757f

**Comment by heinrichI** [comment]
Add PR https://github.com/dmMaze/BallonsTranslator/pull/974

**Comment by bropines** [comment]
> Add PR [#974](https://github.com/dmMaze/BallonsTranslator/pull/974)

Will there be any screenshots of what TV has done there? I'm just on vacation, I want to see the implementation.

**Comment by heinrichI** [comment]
> Will there be any screenshots of what TV has done there? I'm just on vacation, I want to see the implementation.

The only thing I don't know is how to switch the spell check language, now in the spell_check_engine constructor. And scrolling in the SpellCheck panel doesn't work.

https://github.com/user-attachments/assets/56c59193-e8bc-4898-ab3f-00c3b668581a

**Comment by bropines** [comment]
> > Will there be any screenshots of what TV has done there? I'm just on vacation, I want to see the implementation.
> 
> The only thing I don't know is how to switch the spell check language, now in the spell_check_engine constructor. And scrolling in the SpellCheck panel doesn't work.
> 
>  Archie.mp4

Well, it looks good. I asked dmmaze to do the implementation as in regular editors. In theory, I think it's possible to combine these two methods. @dmMaze

**Comment by heinrichI** [comment]
Something else needs to be done. In an unfamiliar language, you need to compare an unknown word with a picture, which can be difficult to understand. You need to guess the meaning, and to do that, you need to show a list of possible corrections along with the translation of these words. But how can this be done in the GUI?
